crates/iota-data-ingestion-core/src/history/reader.rs

```diff
@@ -0,0 +1,313 @@
+    pub async fn sync_manifest_once(&self) -> Result<()> {
+        Self::sync_manifest(self.remote_object_store.clone(), self.manifest.clone()).await?;
+        Ok(())
+    }
+
+    pub async fn get_manifest(&self) -> Manifest {
+        self.manifest.lock().await.clone()
+    }
+
+    /// Copies Manifest from remote store to the given Manifest.
+    async fn sync_manifest(
+        remote_store: Arc<dyn ObjectStoreGetExt>,
+        manifest: Arc<Mutex<Manifest>>,
+    ) -> Result<()> {
+        let new_manifest = read_manifest(remote_store.clone()).await?;
+        let mut locked = manifest.lock().await;
+        *locked = new_manifest;
+        Ok(())
+    }
+
+    /// Resolve the files to fetch for the specified range.
+    ///
+    /// The method retrieves the manifest from the remote store and
+    /// searches for the files that cover the given range of checkpoint
+    /// data.
+    ///
+    /// # Errors
+    ///
+    /// The method fails if the remote store has no data, or if the
+    /// manifest fails to verify.
+    async fn get_files_for_range(
+        &self,
+        checkpoint_range: Range<CheckpointSequenceNumber>,
+    ) -> Result<impl Iterator<Item = FileMetadata>> {
+        let manifest = self.get_manifest().await;
+
+        let latest_available_checkpoint = manifest
+            .next_checkpoint_seq_num()
+            .checked_sub(1)
+            .ok_or_else(|| {
+                IngestionError::HistoryRead("no checkpoint data in the remote store".into())
+            })?;
+
+        if checkpoint_range.start > latest_available_checkpoint {
+            return Err(IngestionError::HistoryRead(format!(
+                "latest available checkpoint is: {latest_available_checkpoint}",
+            )));
+        }
+
+        let files = self.verify_manifest(manifest)?;
+
+        let start_index = match files
+            .binary_search_by_key(&checkpoint_range.start, |s| s.checkpoint_seq_range.start)
+        {
+            Ok(index) => index,
+            Err(index) => index - 1,
+        };
+
+        let end_index = match files
+            .binary_search_by_key(&checkpoint_range.end, |s| s.checkpoint_seq_range.start)
+        {
+            Ok(index) => index,
+            Err(index) => index,
+        };
+
+        Ok(files
+            .into_iter()
+            .enumerate()
+            .filter_map(move |(index, metadata)| {
+                (index >= start_index && index < end_index).then_some(metadata)
+            }))
+    }
+
+    fn spawn_manifest_sync_task<S: ObjectStoreGetExt + Clone>(
+        remote_store: S,
+        manifest: Arc<Mutex<Manifest>>,
+        mut recv: oneshot::Receiver<()>,
+    ) {
+        tokio::task::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(60));
+            loop {
+                tokio::select! {
+                    _ = interval.tick() => {
+                        let new_manifest = read_manifest(remote_store.clone()).await?;
```

Comment:
could the `sync_manifest` function be reused here?

Reply:
1596196633